src/MapViewController.swift

```diff
@@ -417,11 +476,172 @@ open class MapViewController: UIViewController, LocationManagerDelegate, TGRecog
     tgViewController.markerRemove(marker)
     return
   }
+}
 
-  //MARK: - TGRecognizerDelegate
-
+extension MapViewController : TGMapViewDelegate, TGRecognizerDelegate {
+  
+  //MARK : TGMapViewDelegate
+  
+  open func mapView(_ mapView: TGMapViewController, didLoadSceneAsync scene: String) {
+    guard (loadDelegate != nil) && (loadDelegate?.responds(to: #selector(MapLoadDelegate.mapController(_:didLoadSceneAsync:))))! else {
+      return
+    }
+    loadDelegate?.mapController!(self, didLoadSceneAsync: scene)
+  }
+  
+  open func mapViewDidCompleteLoading(_ mapView: TGMapViewController) {
+    guard (loadDelegate != nil) && (loadDelegate?.responds(to: #selector(MapLoadDelegate.mapControllerDidCompleteLoading(_:))))! else {
+      return
+    }
+    loadDelegate?.mapControllerDidCompleteLoading!(self)
+  }
+  
+  open func mapView(_ mapView: TGMapViewController, didSelectFeature feature: [AnyHashable : Any]?, atScreenPosition position: CGPoint) {
+    guard (feature != nil) else {
+      return
+    }
+    guard (selectDelegate != nil) else {
+      return
+    }
+    guard (selectDelegate?.responds(to: #selector(MapSelectDelegate.mapController(_:didSelectFeature:atScreenPosition:))))! else {
+      return
+    }
+    selectDelegate?.mapController!(self, didSelectFeature: feature, atScreenPosition: position)
+  }
+  
+  open func mapView(_ mapView: TGMapViewController, didSelectLabel labelPickResult: TGLabelPickResult?, atScreenPosition position: CGPoint) {
+    guard (labelPickResult != nil) else {
+      return
+    }
+    guard (selectDelegate != nil) else {
+      return
+    }
+    guard (selectDelegate?.responds(to: #selector(MapSelectDelegate.mapController(_:didSelectLabel:atScreenPosition:))))! else {
+      return
+    }
+    selectDelegate?.mapController!(self, didSelectLabel: labelPickResult, atScreenPosition: position)
+  }
+  
+  open func mapView(_ mapView: TGMapViewController, didSelectMarker markerPickResult: TGMarkerPickResult?, atScreenPosition position: CGPoint) {
+    guard (markerPickResult != nil) else {
+      return
+    }
+    guard (selectDelegate != nil) else {
+      return
+    }
+    guard (selectDelegate?.responds(to: #selector(MapSelectDelegate.mapController(_:didSelectMarker:atScreenPosition:))))! else {
+      return
+    }
+    selectDelegate?.mapController!(self, didSelectMarker: markerPickResult, atScreenPosition: position)
+  }
+  
+  //MARK : TGRecognizerDelegate
+  
+  open func mapView(_ view: TGMapViewController, recognizer: UIGestureRecognizer, shouldRecognizeSingleTapGesture location: CGPoint) -> Bool {
+    guard (gestureDelegate != nil) else {
+      return true
+    }
+    guard (gestureDelegate?.responds(to: #selector(MapGestureDelegate.mapController(_:recognizer:shouldRecognizeSingleTapGesture:))))! else {
+      return true
+    }
+    let recognize = (gestureDelegate?.mapController!(self, recognizer: recognizer, shouldRecognizeSingleTapGesture: location))!
+    if !recognize {
```

Comment:
I see. Since it seems that we want to always trigger a pick event whenever a tap event occurs we could simplify it a bit; if we leave the `pick*` calls here without the conditional, and remove it from `didRecognizeSingleTapGesture`.

The order of event would change to: `shouldRecognizeTap` -> `selectionCallback(s)` -> (`didRecognizeTap`).